### PR TITLE
Improve IOBuffer docs

### PR DESCRIPTION
**Pull request by haberdashPI**
Based on the discussion in #55978, I have tried to clarify the documentation of `IOBuffer`.

### CODE REVIEW

**Comment by haberdashPI** [comment]
@vtjnash, pinging you for possible review (??), since you were kind enough to clarify this to me.

File: base/iobuffer.jl [hunk 1]
```diff
@@ -59,6 +59,15 @@ It may take optional keyword arguments:
...
 When `data` is not given, the buffer will be both readable and writable by default.
 
+!!! warn "`data` is invalid after being passed to `IOBuffer`"
+    Once `data` is passed to `IOBuffer`, if `write=true` (or `append=true`) and
+    `maxsize > length(data)`, `IOBuffer` essentially "owns" this data, meaning that
+    existing bindings to `data` should be considered invalid. There are no guarantees on
+    its content at this point. For example, `IOBuffer` may re-allocate the data as required,
+    which may or may not be visible in any outstanding bindings to `array`. There could be
+    unallocated, arbitrary values. Likewise, the ordering of any values written to this
+    array cannot be presumed.
```
Comment (LilithHafner) [inline base/iobuffer.jl:69]: In addition to no guarantees on its content, future mutation of `data` can cause unexpected behavior including data corruption.

By documenting it this way, we guarantee that if write and append are false or maxsize is equal to length(data) then `data` is valid after being passed. I want to make sure that API choice is intentional.

**Comment by vtjnash** [comment]
That seems slightly more an issue with the `String` constructor (and continuing to use the argument afterwards) in v1.11 than with `IOBuffer`. In v1.12, that code is now more safe (due to Memory) and won't segfault (though mutating the Array/Memory after calling String is currently still invalid)

File: base/iobuffer.jl (hunk 1 above)
Comment (vtjnash) [reply base/iobuffer.jl]: That seems slightly more an issue with the `String` constructor (and continuing to use the argument afterwards) in v1.11 than with `IOBuffer`. In v1.12, that code is now more safe (due to Memory) and won't segfault (though mutating the Array/Memory after calling String is currently still invalid)

File: base/iobuffer.jl [hunk 2]
```diff
@@ -55,10 +55,21 @@ It may take optional keyword arguments:
 - `read`, `write`, `append`: restricts operations to the buffer; see `open` for details.
 - `truncate`: truncates the buffer size to zero length.
 - `maxsize`: specifies a size beyond which the buffer may not be grown.
-- `sizehint`: suggests a capacity of the buffer (`data` must implement `sizehint!(data, size)`).
+- `sizehint`: suggests a capacity of the buffer (`data` must implement `sizehint!(data,
+  size)`).
```
Comment (vtjnash) [inline base/iobuffer.jl:59]: We have no strict line length limit
```suggestion
- `sizehint`: suggests a capacity of the buffer (`data` must implement `sizehint!(data, size)`).
```

File: base/iobuffer.jl [hunk 3]
```diff
@@ -55,10 +55,21 @@ It may take optional keyword arguments:
...
 
+!!! warn "`data` is invalid after being passed to `IOBuffer`"
+    Once `data` is passed to `IOBuffer`, it is best to consider `data` invalid; in effect
+    `IOBuffer` "owns" this array from then on. Any mutations to `data` can lead to undefined
+    behavior when reading from the `IOBuffer`. If `write=true` (or `append=true`) and
+    `maxsize > length(data)`, existing bindings to `data` may reference undefined content.
+    For example, `IOBuffer` may re-allocate the data as required, which may or may
+    not be visible in any outstanding bindings to `array`. There could be unallocated,
+    arbitrary values. Likewise, the ordering of any values written to this array cannot
+    be presumed.
```
Comment (vtjnash) [inline base/iobuffer.jl:71]: Some word-smithing suggestions:
```suggestion
!!! warning "Passing `data` as scratch space to `IOBuffer` with `write=true` may give unexpected behavior"
    Once `write` is called on an `IOBuffer`, it is best to consider an
    previous references to `data` invalidated; in effect `IOBuffer` "owns"
    this data until a call to `take!`. Any indirect mutations to `data`
    could lead to undefined behavior by breaking the abstractions expected
    by `IOBuffer`. If `write=true` the IOBuffer may store data at any
    offset leaving behind arbitrary values at other offests. If `maxsize >
    length(data)`, the IOBuffer might re-allocate the data entirely, which
    may or may not be visible in any outstanding bindings to `array`.
```

File: base/iobuffer.jl (hunk 1 above)
Comment (nhz2) [reply base/iobuffer.jl]: Should calling `IOBuffer` on a `Vector{UInt8}` with `write=true` empty the original vector after taking the memory, like the `String` function currently does?

Comment (vtjnash) [reply base/iobuffer.jl]: Sure, if `write=true` and `maxsize > length` that seems a good way to partially encode this restriction in code

**Comment by haberdashPI** [comment]
I'm finished here, I believe (mostly just accepted the suggested changes).

**Review by vtjnash — APPROVED** [review]

**Pull request merged**